### PR TITLE
[Fix #7740] Make `Style/AccessModifierDeclarations` not flag symbol method-name

### LINKED ISSUE #7740: False Positive and Incorrect Error Message for Style/AccessModifierDeclarations
RuboCop 0.80.0 is reporting offences for marking existing methods as private, even when the method is not defined within the scope it is marked private in.
Furthermore, the message produced by the offence is misleading.

--------

## Expected behavior
I expect to be able to mark methods introduced by reflection or otherwise defined outside of the current scope as protected or private.

## Actual behavior
Consider a hypothetical example using a Rails model:
```ruby
ApplicationRecord.connection.create_table(:articles) do |t|
  t.string :name
  t.string :slug
end

class Article < ApplicationRecord
  def name=(value)
    super
    self.slug = name&.parameterize
  end

  private :slug=
end
```

The rubocop offence is reported on the `private :slug=` line:
> `private` should not be inlined in method definitions. (convention:Style/AccessModifierDeclarations)

The reported offence is incorrect since:
* The line is not a method definition, hence the error message is incorrect and misleading.
* The line affects a method that is not defined by application code. It is created by Rails based on the connected database schema, and its natural access modifier is outside of application control.

## Steps to reproduce the problem
Minimum case to reproduce:
```ruby
module Bar
  def bar
    nil
  end
end

class Foo
  include Bar
  private :bar
end
```
The error message is produced:
> `private` should not be inlined in method definitions. (convention:Style/AccessModifierDeclarations)

The line neither defines a method, nor does class `Foo` itself declare the method `bar`.

## RuboCop version
```
0.80.0 (using Parser 2.7.0.2, running on ruby 2.7.0 x86_64-linux)
```

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #7740

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/access_modifier_declarations.rb [hunk 1]
```diff
@@ -65,6 +65,9 @@ def on_send(node)
           return unless node.access_modifier?
           return if node.parent.pair_type?
 
+          argument = node.children.last
```
Comment (bbatsov) [inline lib/rubocop/cop/style/access_modifier_declarations.rb]: I wonder if we can simplify this somehow. Generally I've never been fond of getting all the children in this fashion. I'd rather go with simple destructuring of the node.

File: lib/rubocop/cop/style/access_modifier_declarations.rb [hunk 2]
```diff
@@ -65,6 +65,9 @@ def on_send(node)
...
+          return if argument.is_a?(AST::Node) && argument.sym_type?
```
Comment (bbatsov) [inline lib/rubocop/cop/style/access_modifier_declarations.rb]: Why the explicit type check?

**Comment by bbatsov** [comment]
The cop documentation should definitely be updated to reflect this. 

I think we also need some flag to enable/disable this, as if you're using modifiers in this manner, you're potentially violating grouping/non-grouping. That's why I'd prefer having more control over the behaviour.

**Comment by tejasbubane** [comment]
@bbatsov Fixed to use node-matcher instead of explicit checks. 

Regarding config, do you think the following configs makes sense?
1. group (default) - forbids inline method definition but allows inline symbol
2. strict_group - forbids inline method definition & symbols
3. inline - all modifiers should be inline

Let me know if you have something else in mind.

**Comment by Mihail-K** [comment]
@tejasbubane 
The act of allowing inline-symbol seems tangential to the modes of `group` and `inline`, since those refer to definition style, whereas referencing a Symbol is used for existing declarations.

Maybe it might make more sense to have it be a flag/separate option?
Allowing or forbidding the inline-Symbol behaviour could be applicable regardless of whether the preferred declaration style is inline or group.

**Comment by bbatsov** [comment]
> Maybe it might make more sense to have it be a flag?

Exactly. This shouldn't be a style, it should a flag that affects both styles - something like `AllowModifiersOnSymbols` or whatever. A better name would be welcome. :-)

**Comment by tejasbubane** [comment]
@bbatsov I added config option to `AllowModifiersOnSymbols`.

File: lib/rubocop/cop/style/access_modifier_declarations.rb [hunk 3]
```diff
@@ -48,6 +48,24 @@ module Style
       #     private def baz; end
       #
       #   end
+      # @example AllowModifiersOnSymbols: true
```
Comment (bbatsov) [inline lib/rubocop/cop/style/access_modifier_declarations.rb]: The formatting here seems off.

File: config/default.yml [hunk 4]
```diff
@@ -2202,6 +2202,7 @@ Style/AccessModifierDeclarations:
   SupportedStyles:
     - inline
     - group
+  AllowModifiersOnSymbols: true
```
Comment (bbatsov) [inline config/default.yml]: You'll have to update `VersionChanged` as well.

File: lib/rubocop/cop/style/access_modifier_declarations.rb (hunk 3 above)
Comment (bbatsov) [reply lib/rubocop/cop/style/access_modifier_declarations.rb]: Hmm, I see now it's messed up in the entire file. I guess you just stuck to the existing examples.

File: config/default.yml (hunk 4 above)
Comment (tejasbubane) [reply config/default.yml]: Added version `0.81`

File: lib/rubocop/cop/style/access_modifier_declarations.rb (hunk 3 above)
Comment (tejasbubane) [reply lib/rubocop/cop/style/access_modifier_declarations.rb]: Do you mean the empty line before `@example`?

Comment (bbatsov) [reply lib/rubocop/cop/style/access_modifier_declarations.rb]: That and the empty lines between `bad/good`. Just compare this to other source files and you'll see what I mean.

**Comment by bbatsov** [comment]
I'd also extend the cop documentation to mention that it covers only definitions and applications of visibility methods to symbols are controlled by the new config.

File: lib/rubocop/cop/style/access_modifier_declarations.rb (hunk 3 above)
Comment (tejasbubane) [reply lib/rubocop/cop/style/access_modifier_declarations.rb]: Fixed.

**Comment by tejasbubane** [comment]
> I'd also extend the cop documentation to mention that it covers only definitions and applications of visibility methods to symbols are controlled by the new config.

Added.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**